### PR TITLE
[confmap] Store original string in confmap.Conf

### LINKED ISSUE #10405: Error loading TLS certificates from ENV variable
**Describe the bug**
When attempting to configure TLS using environment variables for the contents of the certs there is an error parsing the value of the certificates.

**Steps to reproduce**
Use the following configuration file:

```
---
receivers:
  otlp:
    protocols:
      http:
        endpoint: 0.0.0.0:4318
processors:
  batch:

exporters:
  kafka:
    auth:
      tls:
        # DOES NOT WORK: Contents of the cert in the env variable
        ca_pem: ${env:CA_PEM}

        # WORKS
        # ca_pem: |
        #   -----BEGIN CERTIFICATE-----
        #   MIICUTCCAfugAwIBAgIBADANBgkqhkiG9w0BAQQFADBXMQswCQYDVQQGEwJDTjEL
        #   MAkGA1UECBMCUE4xCzAJBgNVBAcTAkNOMQswCQYDVQQKEwJPTjELMAkGA1UECxMC
        #   VU4xFDASBgNVBAMTC0hlcm9uZyBZYW5nMB4XDTA1MDcxNTIxMTk0N1oXDTA1MDgx
        #   NDIxMTk0N1owVzELMAkGA1UEBhMCQ04xCzAJBgNVBAgTAlBOMQswCQYDVQQHEwJD
        #   TjELMAkGA1UEChMCT04xCzAJBgNVBAsTAlVOMRQwEgYDVQQDEwtIZXJvbmcgWWFu
        #   ZzBcMA0GCSqGSIb3DQEBAQUAA0sAMEgCQQCp5hnG7ogBhtlynpOS21cBewKE/B7j
        #   V14qeyslnr26xZUsSVko36ZnhiaO/zbMOoRcKK9vEcgMtcLFuQTWDl3RAgMBAAGj
        #   gbEwga4wHQYDVR0OBBYEFFXI70krXeQDxZgbaCQoR4jUDncEMH8GA1UdIwR4MHaA
        #   FFXI70krXeQDxZgbaCQoR4jUDncEoVukWTBXMQswCQYDVQQGEwJDTjELMAkGA1UE
        #   CBMCUE4xCzAJBgNVBAcTAkNOMQswCQYDVQQKEwJPTjELMAkGA1UECxMCVU4xFDAS
        #   BgNVBAMTC0hlcm9uZyBZYW5nggEAMAwGA1UdEwQFMAMBAf8wDQYJKoZIhvcNAQEE
        #   BQADQQA/ugzBrjjK9jcWnDVfGHlk3icNRq0oV7Ri32z/+HQX67aRfgZu7KWdI+Ju
        #   Wm7DCfrPNGVwFWUQOmsPue9rZBgO
        #   -----END CERTIFICATE-----

        # WORKS: Above contents stored to a file
        #ca_file: ca.pem

        insecure_skip_verify: true

service:
  pipelines:
    logs:
      receivers: [otlp]
      processors: [batch]
      exporters: [kafka]
```

Create a `ca.pem` file with the content:

```
-----BEGIN CERTIFICATE-----
MIICUTCCAfugAwIBAgIBADANBgkqhkiG9w0BAQQFADBXMQswCQYDVQQGEwJDTjEL
MAkGA1UECBMCUE4xCzAJBgNVBAcTAkNOMQswCQYDVQQKEwJPTjELMAkGA1UECxMC
VU4xFDASBgNVBAMTC0hlcm9uZyBZYW5nMB4XDTA1MDcxNTIxMTk0N1oXDTA1MDgx
NDIxMTk0N1owVzELMAkGA1UEBhMCQ04xCzAJBgNVBAgTAlBOMQswCQYDVQQHEwJD
TjELMAkGA1UEChMCT04xCzAJBgNVBAsTAlVOMRQwEgYDVQQDEwtIZXJvbmcgWWFu
ZzBcMA0GCSqGSIb3DQEBAQUAA0sAMEgCQQCp5hnG7ogBhtlynpOS21cBewKE/B7j
V14qeyslnr26xZUsSVko36ZnhiaO/zbMOoRcKK9vEcgMtcLFuQTWDl3RAgMBAAGj
gbEwga4wHQYDVR0OBBYEFFXI70krXeQDxZgbaCQoR4jUDncEMH8GA1UdIwR4MHaA
FFXI70krXeQDxZgbaCQoR4jUDncEoVukWTBXMQswCQYDVQQGEwJDTjELMAkGA1UE
CBMCUE4xCzAJBgNVBAcTAkNOMQswCQYDVQQKEwJPTjELMAkGA1UECxMCVU4xFDAS
BgNVBAMTC0hlcm9uZyBZYW5nggEAMAwGA1UdEwQFMAMBAf8wDQYJKoZIhvcNAQEE
BQADQQA/ugzBrjjK9jcWnDVfGHlk3icNRq0oV7Ri32z/+HQX67aRfgZu7KWdI+Ju
Wm7DCfrPNGVwFWUQOmsPue9rZBgO
-----END CERTIFICATE-----
```

Set env variable to the content of the cert:
```
export CA_PEM=$(cat ca.pem)
```

Run otel collector with the configuration file:

`otelcol --config config.yml`

**What did you expect to see?**
I expected that when `ca_pem: ${env:CA_PEM}` was specified that otel collector would parse the cert.

**What did you see instead?**
When `ca_pem: ${env:CA_PEM}` is specified, I get the following error:

```
Error: cannot start pipelines: error loading tls config: failed to load TLS config: failed to load CA CertPool PEM: failed to parse cert
2024/06/13 18:05:57 collector server run finished with error: cannot start pipelines: error loading tls config: failed to load TLS config: failed to load CA CertPool PEM: failed to parse cert
```

When I comment out that configuration and use either `ca_pem: | CONTENT` or `ca_file: ca.pem` the above error is not present and the collector starts up.

**What version did you use?**
Version: 0.102.1

**What config did you use?**
The above configuration file.

**Environment**
Docker image: `golang:1.22`

`otelcol` was fetched from Github and unpacked within the docker image manually.

```
OTEL_COLLECTOR_VERSION=0.102.1
OTEL_COLLECTOR_ARTIFACT=otelcol_${OTEL_COLLECTOR_VERSION}_linux_amd64.tar.gz
OTEL_COLLECTOR_DEST=.
wget https://github.com/open-telemetry/opentelemetry-collector-releases/releases/download/v$OTEL_COLLECTOR_VERSION/$OTEL_COLLECTOR_ARTIFACT
tar -xzvf $OTEL_COLLECTOR_ARTIFACT -C $OTEL_COLLECTOR_DEST otelcol
```

**Additional context**

### LINKED ISSUE #10605: Collector fails to expand env vars when the value is an IPv6 address with trailing double colon
**Describe the bug**

On startup, the collector fails to expand env vars when the value is an IPv6 address with trailing double colon. It treats the value as a map instead of a string.

**Steps to reproduce**

In my environment, I see this when the collector Pod's IPv6 address has a trailing double colon. Given `MY_POD_IP=1111:1111:1111:1111:1111::`, this fails to expand:

```yaml
config:
  receivers:
    zipkin:
      endpoint: '[${env:MY_POD_IP}]:9411'
```

If you don't have an environment where you can get a IPv6 Pod, you can also reproduce like so:

```yaml
extraEnvs:
- name: MY_MOCK_IP
  value: '1111:1111:1111:1111:1111::'
# ...
config:
  receivers:
    zipkin:
      endpoint: '[${env:MY_MOCK_IP}]:9411'
```

**What did you expect to see?**

The collector start successfully with its config expanded, regardless of whether the Pod's IPv6 address has a leading or trailing double colon.

**What did you see instead?**

```shell
Error: failed to resolve config: cannot resolve the configuration: expanding ${env:MY_POD_IP}: expected convertable to string value type, got map["1111:1111:1111:1111:1111:":<nil>](map[string]interface {})
2024/07/12 17:54:20 collector server run finished with error: Error: failed to resolve config: cannot resolve the configuration: expanding ${env:MY_POD_IP}: expected convertable to string value type, got map["1111:1111:1111:1111:1111:":<nil>](map[string]interface {})
```

**What version did you use?**

`0.104.0`

**What config did you use?**

```yaml
config:
  receivers:
    zipkin:
      endpoint: '[${env:MY_POD_IP}]:9411'
```

**Environment**

Kubernetes Version: `1.28.11-eks-db838b0`
Helm Chart Version: `0.97.1`
Docker Image Version: `otel/opentelemetry-collector-contrib:0.104.0`

**Additional context**

Looks like this may have been partially resolved with #10343

### LINKED ISSUE #10659: Issue enforcing quoted values in config
**Describe the bug**

When providing a configuration that uses the `env` provider, it does not preserve the quotes around the `"${env:...}"`.

This means when a complaint time RFC string is supplied, it gets converted into a `time.Time` value instead of the expected string, which then stops the config from being valid.

**Steps to reproduce**

1. Set an env var with a timestamp value
  a. `export BUILT_TS="2006-01-02T15:04:05Z07:00"`
1. Reference var in configuration.
  a. `value : "${env:BUILT_TS}"`
1. Have collector run with configuration using reference variable.

**What did you expect to see?**
Considering the provider is in quotes, it should not strip those values

**What did you see instead?**
Upon running the collector with the provider configuration, the error is show as:
```
Error: failed to get config: cannot resolve the configuration: unsupported type=time.Time for retrieved config, ensure that values are wrapped in quotes
```

**What version did you use?**
v0.105.0

**What config did you use?**

```yaml
processors:
  resource/static:
    attributes:
    - key: build.version
      action: insert
      value: "${BUILD_VERSION}"
```

**Environment**
Linux, go v1.23.4

**Additional context**
I've tried a few different combinations, but this currently works in `v0.103.0` but upgrading to `v0.105.0` surfaced this. 

I've also tried:
```yaml
\"${env:var}\" # Didn't work with error: retrieved value does not have unambiguous string representation: <nil>
"${env:var}"  # moved to calling the env provider directly, same error as reported
```

### PR DESCRIPTION
**Pull request by mx-psi**
#### Description

- Adds new `expandedValue` struct that holds the original string representation if available for values resolved from a provider.
- Removes any mention of `expandedValue` in the public API by adding a `sanitize` step before returning any `Get`s or `ToStringMap`s.
- Adds new decoding hook that checks if the target field is of `string` type and uses the string representation in that case.

#### Link to tracking issue

Fixes #10605, Fixes #10405, Fixes #10659

#### Testing

This changes the behavior in some cases, I update the test cases.

#### Documentation

| ENV value                  | ${ENV} before unification  | ${ENV} in v0.105.0 (also ${env:ENV} before unification) | Value after this PR        |
|----------------------------|----------------------------|---------------------------------------------------------|----------------------------|
| foo\nbar                   | foo\nbar                   | foo bar                                                 | foo\nbar                   |
| 1111:1111:1111:1111:1111:: | 1111:1111:1111:1111:1111:: | **Error**                                               | 1111:1111:1111:1111:1111:: |
| "0123"                     | "0123"                     | 0123                                                    | "0123"                     |

### CODE REVIEW

**Comment by mx-psi** [comment]
I still need to think about some edge cases and add more tests, but the overall approach should be clear now and can be reviewed

File: confmap/expand.go [hunk 1]
```diff
@@ -43,6 +43,20 @@ func (mr *Resolver) expandValueRecursively(ctx context.Context, value any) (any,
...
 	switch v := value.(type) {
+	case expandedValue:
+		expanded, changed, err := mr.expandValue(ctx, v.Value)
+		if err != nil {
+			return nil, false, err
+		}
+
+		if _, ok := expanded.(expandedValue); ok {
+			return expanded, changed, nil
+		}
```
Comment (mx-psi) [inline confmap/expand.go:54]: This part needs a test where we have `field: ${env:VAL1}`, `VAL1=${env:VAL2}`, `VAL2=1234`

**Review by codeboten — COMMENTED** [review]
I think this approach makes sense to address the issue, there are some lint errors that need addressing

**Review by TylerHelmuth — COMMENTED** [review]
I agree that this looks like the right approach to me as it lets users continue to set yaml via env vars but also use exact inline env var values in strings.

**Comment by mx-psi** [comment]
@MovieStoreGuy it would really help if you can try this one out and confirm that it fixes your issue in an environment as close as possible to the one where you found it

**Review by TylerHelmuth — APPROVED** [review]

**Comment by mx-psi** [comment]
Final call for reviews, I want to merge this by Thursday EU morning cc @open-telemetry/collector-approvers
